src/xdsl/rewriter.py

```diff
@@ -0,0 +1,199 @@
+import inspect
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from typing import List, Optional, Callable, Dict
+
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import Operation, SSAValue, Region, Block
+
+
+@dataclass(init=False)
+class RewriteAction:
+    """
+    Action that a single rewrite may execute.
+    A rewrite always delete the matched operation, and replace it with new operations.
+    The matched operation results are replaced with new ones.
+    """
+
+    new_ops: List[Operation]
+    """New operations that replace the one matched."""
+
+    new_results: List[SSAValue]
+    """SSA values that replace the matched operation results."""
+    def __init__(self, new_ops: List[Operation],
+                 new_results: Optional[List[SSAValue]]):
+        self.new_ops = new_ops
+        if new_results is not None:
+            self.new_results = new_results
+            return
+
+        if len(new_ops) == 0:
+            self.new_results = []
+        else:
+            self.new_results = new_ops[-1].results
```

Comment:
I'm not sure if we should allow so many different options. Forcing a user to explicitly pass the list of new results might be acceptable. Making `new_results` default value an empty list might also be an option.

Reply:
An empty list as default for new_results seems sensible for me as well.

Reply:
Well, my intuition is that for the big majority of the cases, we are replacing an operation with multiple others and mapping the last operation results on a new one, or removing an operation.
Making this an empty list by default would make ask users to explicitly add the last operation results almost every time.
What I could also do is write another constructor

Reply:
How can you write another constructor? Python doesn't support method overloading. 

Maybe one could write static factory methods, but I'm not sure what's the "pythonic way" here.

Reply:
Yes, here I mean another static factory method!

Reply:
Should be added now